# 1.异常概述

在程序运行的过程中,经常会遇到各种各样的错误,这些错误统称为异常,

# 2.python中常见的异常和描述

| AssertionError        | 断言语句（assert）失败                                       |
| --------------------- | ------------------------------------------------------------ |
| AttributeError        | 尝试访问一个对象没有的属性，比如foo.x，但是foo没有属性x      |
| EOFError              | 用户输入文件末尾标志EOF（Ctrl+d）,input读取到EOF却没有接受任何数据 |
| FloatingPointError    | 浮点计算错误                                                 |
| GeneratorExit         | generator.close()方法被调用的时候                            |
| ImportError           | 导入模块失败的时候                                           |
| IndexError            | 索引超出序列的范围                                           |
| KeyError              | 字典中查找一个不存在的关键字                                 |
| KeyboardInterrupt     | 用户输入中断键（Ctrl+c）                                     |
| MemoryError           | 内存溢出（可通过删除对象释放内存）                           |
| NameError             | 尝试访问一个不存在的变量                                     |
| NotImplementedError   | 尚未实现的方法                                               |
| OSError               | 操作系统产生的异常（例如打开一个不存在的文件）               |
| OverflowError         | 数值运算超出最大限制                                         |
| ReferenceError        | 弱引用（weak reference）试图访问一个已经被垃圾回收机制回收了的对象 |
| RuntimeError          | 一般的运行时错误                                             |
| StopIteration         | 迭代器没有更多的值                                           |
| SyntaxError           | Python的语法错误                                             |
| IndentationError      | 缩进错误                                                     |
| TabError              | Tab和空格混合使用                                            |
| SystemError           | Python编译器系统错误                                         |
| SystemExit            | Python编译器进程被关闭                                       |
| TypeError             | 不同类型间的无效操作                                         |
| UnboundLocalError     | 访问一个未初始化的本地变量（NameError的子类）                |
| UnicodeError          | Unicode相关的错误（ValueError的子类）                        |
| UnicodeEncodeError    | Unicode编码时的错误（UnicodeError的子类）                    |
| UnicodeDecodeError    | Unicode解码时的错误（UnicodeError的子类）                    |
| UnicodeTranslateError | Unicode转换时的错误（UnicodeError的子类）                    |
| ValueError            | 传入无效的参数                                               |
| ZeroDivisi            | 除数为零                                                     |

# 3.异常处理语句

1. try   except  异常类型

```python
#语法
#try:
    #可能存在异常的语句
#except [异常类型as为当前异常指定的别名]
    #进行异常处理的代码
def dis(   ):
    apple=input('输入苹果的数量:')
    children=input('输入孩子的数量')
    if(apple/children>1):
        print('苹果绰绰有余')
    elif(apple/children==1):
        print('苹果刚刚好')
    else:
        print('苹果不够分')

try:
 dis()
except ZeroDivisionError:
    print('出错了,苹果不能为0')
```

2.try  except  异常类型   else    在else语句中,是没有发生异常时执行的程序

3.try  except  异常类型   finally 在finally语句中是最终一定会执行的程序

# 4.raise抛出异常

```python
if  apple<children:
   raise('苹果太少了不够分')
```

